### PR TITLE
Import relayOperationGenericsRule

### PR DESCRIPTION
**Pull request by ashfurrow**
This is a work-in-progress PR to import [relayOperationGenericsRule.js](https://github.com/artsy/reaction/blob/48f7446cee9f38ec4f1c153c1f105bdbee2913df/tslint/relayOperationGenericsRule.js) into this project, including a transition to TypeScript proper instead of `@ts-check`. It's a WIP because I ran out of time this morning. Next steps are to finish it, link it from Reaction, and get Reaction consuming the rule from a module. Then we can think about tests, too, and then add [this](https://github.com/relayjs/eslint-plugin-relay/blob/master/src/rule-unused-fields.js) and other eslint-plugin-relay rules to this project. And then publish to NPM. Baby steps.

See also: https://github.com/artsy/yeoman-generator-artsy/pull/10

/cc @alloy

### CODE REVIEW

**Comment by ashfurrow** [comment]
Aww jeez

File: package.json [hunk 1]
```diff
@@ -48,5 +49,8 @@
       "yarn prettier --write",
       "git add"
     ]
+  },
+  "dependencies": {
+    "graphql": "^14.0.2"
```
Comment (alloy) [inline package.json]: It’s customary to make these types of dependencies peer ones, although I think that’s more important when it’s a dep that will be used at runtime to ensure the app has only 1 version of it.

File: src/relayOperationGenericsRule.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      this.addFailureAtNode(taggedTemplate, "expected a graphql`…` tagged-template")
+    }
+  }
+
+  /**
+   * @param {number} start
+   * @param {number} width
+   * @param {string} replacement
+   * @param {string} operationName
+   * @returns {Lint.Replacement[]}
```
Comment (alloy) [inline src/relayOperationGenericsRule.ts]: I believe these type annotations can be removed when using TS, which already has these in its function signature.

File: src/relayOperationGenericsRule.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
+import { BREAK, parse, visit } from "graphql"
```
Comment (alloy) [inline src/relayOperationGenericsRule.ts]: Might be worth it to import this `* as graphql` as these symbols could be slightly confusing when used together with the tsc API

**Comment by alloy** [comment]
Ah yes! Nice 👌

**Comment by alloy** [comment]
Btw, I’d appreciate it if you could make me the author of the initial commit, even if just for knowing who to talk to if there’s ever any questions 🙏

**Comment by alloy** [comment]
Let‘s move it to the relay-tools org before cutting the first release. I’ll invite you to the org now.

**Comment by ashfurrow** [comment]
Sure, sounds good 👍 Thanks!

File: src/relayOperationGenericsRule.ts (hunk 2 above)
Comment (ashfurrow) [reply src/relayOperationGenericsRule.ts]: Yup, I've been working my way down the file and switching them over 👍

File: package.json (hunk 1 above)
Comment (ashfurrow) [reply package.json]: Ah, interesting! It took until now for me to get the distinction, [this blog post](https://yarnpkg.com/blog/2018/04/18/dependencies-done-right/) was helpful.

Comment (alloy) [reply package.json]: It’s an art!

**Comment by alloy** [comment]
Thanks!

File: package.json [hunk 4]
```diff
@@ -17,7 +20,9 @@
     "ts-jest": "^23.10.1",
     "ts-node": "^7.0.1",
     "tslint": "^5.11.0",
-    "typescript": "^3.0.3"
+    "tslint-config-prettier": "^1.15.0",
+    "typescript": "^3.0.3",
+    "graphql": "^14.0.2"
```
Comment (ashfurrow) [inline package.json]: I had to put this in `devDependencies` because peer deps aren't installed by default.

**Comment by ashfurrow** [comment]
Okay, this just needs testing now. Before we publish, we can link locally on reaction and make sure it will work. I'm expecting to have to faff with TSLint to get it to load rules from node modules.

**Comment by ashfurrow** [comment]
I've found this blog post helpful for getting the package set up and linked from Reaction: https://palantir.github.io/tslint/2016/03/31/sharable-configurations-rules.html

Also, it's not possible to link against a package that has no published version ([see this issue](https://github.com/yarnpkg/yarn/issues/2611)) so I added the following to Reaction's `package.json` to test:

```json
"tslint-plugin-relay": "link:../tslint-plugin-relay"
```

Which sets up the symlink on `yarn install`.

**Comment by ashfurrow** [comment]
Okay, the blog post I linked to is a bit outdated since it doesn't permit TypeScript rules. I've modified how we compile to point to the `rules` directory, for consistency. 

It's not working when linked locally, but I think it might be a problem with the module resolution that our symlink is not adhering too. I'm going to merge, move the repo to the relay-tools org, then publish a v0.0.1 so we can test on Reaction.

**Comment by ashfurrow** [comment]
Woo! https://www.npmjs.com/package/tslint-plugin-relay

Okay, linking works. I looked and there were two problems:

- `tslint-base.json` gets copied to `build` on `yarn build`, but references the `./build` directory _from_ there, causing path problems. 
- Reaction's `tslint.json` file needs to extend `"tslint-plugin-relay"`, not `"tslint-plugin-relay/tslint-config"` as described in the blog post above.

I'll fix both, verify this is working, then PR to Reaction 👍

**Comment by alloy** [comment]
Woohoo! Nice job, @ashfurrow 👏

**Pull request merged**